addons/web/models/statusbar_time_mixin.py

```diff
@@ -0,0 +1,72 @@
+from odoo import api, fields, models
+
+
+class StatusbarTimeMixin(models.AbstractModel):
+    _name = "statusbar.time.mixin"
+    _description = "Mixin to compute the time a record spent in each stage"
+    _stage_field = "stage_id"
+
+    json_statustime = fields.Json(
+        compute="_compute_json_statustime", store=True, help="JSON string that maps stage ids to ms", default={"data": {}})
+    previous_status_id = fields.Integer()
+    has_previous_status_id = fields.Boolean(default=False)
+    previous_status_time = fields.Datetime(
+        help="Time when the record was in its previous stage", default=lambda s: fields.Datetime.now())
+
+    @api.depends(lambda self: [self._stage_field])
+    def _compute_json_statustime(self):
+        """
+        Computes json_statustime, a Json field stored as {"data": { <stage_id (str)>: <time_spent_in_seconds (int)> }}
+
+            e.g. {"data": {"1": 1230, "3": 2220, "5": 14}}
+        """
+        non_virtual_records = self.filtered(lambda r: not isinstance(r.id, models.NewId))
+        records_with_previous_stage = non_virtual_records.filtered(lambda r: r.has_previous_status_id)
+
+        # Get all stage_ids of the batch (all the stages with a time + previous stage + current stage)
+        all_stage_ids = set()
+        for record in records_with_previous_stage:
+            if 'data' in record.json_statustime:
+                all_stage_ids.update(int(stage_id) for stage_id in record.json_statustime['data'].keys())
+            all_stage_ids.add(record.previous_status_id)
+
+        if records_with_previous_stage:
+            all_stage_ids.update([current_stage.id for current_stage in self.mapped(self._stage_field)])
+            all_stages_set_sorted = self.env[self._fields[self._stage_field]._related_comodel_name].browse(all_stage_ids).sorted()
+            stage_ids_indexes = {stage_id.id: index for (index, stage_id) in enumerate(all_stages_set_sorted)}
+
+        for record in records_with_previous_stage:
+            if record[record._stage_field]:
+                record.json_statustime = {"data": record._get_json_ms_in_stages(stage_ids_indexes)}
+            else:
+                record.json_statustime = {"data": {}}
```

Comment:
Please avoid storing useless data (pun unintended) in the database.
```suggestion
            if record[record._stage_field]:
                record.json_statustime = record._get_json_ms_in_stages(stage_ids_indexes)
            else:
                record.json_statustime = False
```